### PR TITLE
Added advanced legend module, see #11309.

### PR DESCRIPTION
**Pull request by kamilkulig**
**Demos:**

- Basic: http://jsfiddle.net/BlackLabel/j6ek3vw1/
- Rich: http://jsfiddle.net/BlackLabel/Lcj06h2y/

**Things to do:**

- [x] ~~add docs for `legend` option in series and color axis options~~
- [x] ~~add support for multiple bubble legends in one chart~~ (there can be only one bubble legend per chart - consulted with @ppotaczek)
- [x] ~~refactor the code in places where indicated by `// TODO: ...` comments~~ (all essential TODOs resolved)

Nice to have in the future:
- [x] ~~allow to import advanced legend module without loading color axis and bubble legend modules beforehand~~ (However, color axis and bubble legend will appear in the advanced legend only if `modules/advanced-legend` is loaded **after** `modules/heatmap` & `highcharts-more`.)
- [ ] allow referring to a legend by index

### CODE REVIEW

**Comment by sebastianbochan** [comment]
I tried to reproduce the problem with test without success.

Safari: 12.1.1
MacOSX: 10.14.5 Mojave 
 
1. http://jsfiddle.net/BlackLabel/vtj0ec4x/ - works properly
2. karma - all tests passed
3. `/samples/unit-tests/responsive/responsive/` - test passed

Does anyone have the specific version of Safari?

File: js/parts/ColumnSeries.js [hunk 1]
```diff
@@ -577,7 +577,9 @@ seriesType('column', 'line',
         Series.prototype.translate.apply(series);
         // Record the new values
         series.points.forEach(function (point) {
-            var yBottom = pick(point.yBottom, translatedThreshold), safeDistance = 999 + Math.abs(yBottom), pointWidth = seriesPointWidth, 
+            var yBottom = pick(point.yBottom, translatedThreshold, 0 // Handle the situation when the plot area is not
+            // visible because of multiple legends.
+            ), safeDistance = 999 + Math.abs(yBottom), pointWidth = seriesPointWidth,
```
Comment (sebastianbochan) [inline js/parts/ColumnSeries.js]: What is the reason of using 999 ? Why i.e 1294?

Comment (kamilkulig) [reply js/parts/ColumnSeries.js]: That line of code is not written by me. I've just added another argument to `pick` function.

**Comment by TorsteinHonsi** [comment]
Please merge the master into this PR so we can fix the conflicts and get an overview over the changes.

I also see there is a lot of diff on core components, so when we have merged in the master we will see how much is refactoring and how much are additions.

**Comment by bre1470** [comment]
@kamilkulig If it helps, I could try to solve the merge conflicts.

**Comment by kamilkulig** [comment]
@bre1470 thanks, that'd be great :)

**Pull request closed without merging**